Add Loris Chiocca's IBM Music Feature Card (IMFC) patch

kcgen opened the pull request:
This PR imports Loris Chiocca's IBM Music Feature Card emulator - a rare and expensive audio card (even for the 80s!) made by IBM that was primarily used by the early Sierra EGA (SCI0) games.

The card runs a Z80 processor, and uses a Yamaha FM chip to generate instrument banks, but can also have custom FM banks uploaded.

LGR just released a video about the Music Feature card: https://www.youtube.com/watch?v=lyIBfjsTZcQ

Sierra's musicians did an OK job with it (look for the `IMF.DRV` file in their games); it's "sound" like OPL'ish MIDI.  NerdlyPleasures has a thorough article about it: http://nerdlypleasures.blogspot.com/2015/02/the-ibm-music-feature-card-overpriced.html

Silpheed's composers probably did it the most justice (don't mind the distracted piloting):

https://user-images.githubusercontent.com/1557255/228120160-cdb2965b-e92c-4490-a383-d73a0ccc4aec.mp4

---

All the add-on commits primarily address CI gates and getting it tied into Staging's slightly changed mixer, IO binds, and more direct init approach.

ThomasEricB commented:
Nice music! Definitely a great addition!

johnnovak commented on include/hardware.h:
```diff
@@ -70,6 +70,9 @@ void CAPTURE_VideoStop();
 void GUS_AddConfigSection(const config_ptr_t &conf);
 
+// IBM Music Feature Card configuration and initialization
+void IMF_AddConfigSection(const config_ptr_t& conf);
```
I think it's abbreviated as `IMFC` or `MFC`. I'd prefer `IMFC`
https://en.wikipedia.org/wiki/IBM_Music_Feature_Card

johnnovak commented on src/dosbox.cpp:
```diff
@@ -846,6 +845,9 @@ void DOSBOX_Init()
 	// Configure Gravis UltraSound emulation
 	GUS_AddConfigSection(control);
 
+	// Configure the IBM Music Feature emulation
```
"IBM Music Feature Card"
In general we should not allow card in the name (so no "Sound Blaster card", "Gravis Ultrasound card"), but in this case "card" is part of the official name and abbreviation.

kcgen replied:
![2023-03-27_21-55](https://user-images.githubusercontent.com/1557255/228132269-bad26b56-f69d-444f-9746-7a33553c325c.png)

kcgen replied:
https://ibm-pc.org/manuals/ibm/options/IMFC_Manual.pdf

IBM's title for it is "IBM PC Music Feature", but lots of installers tacked on the "card".  :man_shrugging: 

The files were originally named 'imfc.cpp' - so it's not a big deal going back to IMFC (there's a commit where I actually switch over to dropping the card).

ThomasEricB replied:
> IBM's title for it is "IBM PC Music Feature", but lots of installers tacked on the "card". 🤷‍♂️

Are we here emulating both the card and the MIDI Adapter Unit?

ThomasEricB replied:
If the code only tackles emulating the music card portion, I would still recommend @johnnovak 's suggestion.

kcgen replied:
Sierra driver also calls it IMF:

![2023-03-27_22-24](https://user-images.githubusercontent.com/1557255/228136561-744f2bbf-7ab8-49ee-a7d9-d32b8a24faf0.png)

ThomasEricB replied:
After hearing LGR's explanation, I think I agree even more with @johnnovak here. It should be "IBM Music Feature Card" instead of "IBM Music Feature", because I believe the "IBM Music Feature " encompasses the MIDI Adapter Unit as well!

kcgen replied:
That's a great point, @ThomasEricB - and that makes a lot of sense.

And you're right: this is emulating the IMF Card, and not the external MIDI breakout box (which would require hooking up an FB-01 device or using an FB-01 emulator).

Will move it back to IMFC.

johnnovak commented on src/hardware/meson.build:
```diff
@@ -33,6 +33,7 @@ libhardware_sources = files(
     'hardware.cpp',
     'ide.cpp',
     'innovation.cpp',
+    'imf.cpp',
```
imfc.cpp

johnnovak replied:
and imfc_rom.c, etc.

kcgen replied:
Reverted; good call.

johnnovak commented on src/hardware/imf_rom.c:
```diff
@@ -0,0 +1,1082 @@
+ *  Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+ */
+
+// Read-only memory tables for the IBM Music Feature
```
"IBM Music Feature Card"

kcgen replied:
Reverted; good call.

johnnovak commented on src/hardware/imf_rom.c:
```diff
@@ -0,0 +1,1082 @@
+
+// Read-only memory tables for the IBM Music Feature
+
+static constexpr uint8_t m_romPresetConfiguration16Binary[160] = {
```
Get rid of the `m_` prefixes? Should be just `RomPresetConfiguration16Binary`, etc.

ThomasEricB replied:
@johnnovak Is that Hungarian notation?

Actual question, but it's funny to ask you that LOL

johnnovak replied:
Yeah, I think so... 😄 It's called like that because it was introduced by [Charles Simonyi](https://en.wikipedia.org/wiki/Charles_Simonyi), who's Hungarian.

kcgen replied:
(marking resolved because not renaming variables)

johnnovak commented on src/hardware/imf.cpp:
```diff
@@ -0,0 +1,13422 @@
+#define IMF_VERBOSE_LOGGING 0
+
+constexpr uint8_t AVAILABLE_MIDI_CHANNELS = 16;
+constexpr uint8_t AVAILABLE_INSTRUMENTS   = 8;
```
UpperCamelCase

johnnovak commented on src/hardware/imf.cpp:
```diff
@@ -0,0 +1,13422 @@
+	PACKET_STATE_3E = 0x3E,
+	PACKET_STATE_3F = 0x3F,
+	PACKET_STATE_40 = 0x40,
+};
```
All enum names and values should be UpperCamelCase (subject to much time you want to spend on this...)

kcgen replied:
(marking resolved because not renaming variables)

johnnovak commented on src/hardware/imf.cpp:
```diff
@@ -0,0 +1,13422 @@
+	volatile bool m_debug                       = false;
+	// volatile BufferDataType m_buffer[2048]; // maximum that is used in
+	// the IMF code
+	volatile BufferDataType m_buffer[0x2000] = {}; // FIXME
```
Getting rid of the `m_` prefixes should be pretty easy.

kcgen replied:
(marking resolved because not renaming variables)

johnnovak commented on src/hardware/imf.cpp:
```diff
@@ -0,0 +1,13422 @@
+	inline void setDataCleared()
+	{
+		flags.setDataCleared();
+	}
```
How about using UpperCamelCase for method names for consistency?

kcgen replied:
(marking resolved because not renaming variables)

johnnovak commented on src/hardware/imf.cpp:
```diff
@@ -0,0 +1,13422 @@
+	unsigned int getLastReadByteIndex()
+	{
+		return lastReadByteIndex;
+	}
```
const

johnnovak commented on src/hardware/imf.cpp:
```diff
@@ -0,0 +1,13422 @@
+	unsigned int getIndexForNextWriteByte()
+	{
+		return indexForNextWriteByte;
+	}
```
const

johnnovak commented on src/hardware/imf.cpp:
```diff
@@ -0,0 +1,13422 @@
+	{
+		return lastReadByteIndex ==
+		       ((indexForNextWriteByte + 1) % m_bufferSize);
+	}
```
const

johnnovak commented on src/hardware/imf.cpp:
```diff
@@ -0,0 +1,13422 @@
+	BufferDataType peekData()
+	{
+		return m_buffer[lastReadByteIndex];
+	}
```
const

johnnovak commented on src/hardware/imf.cpp:
```diff
@@ -0,0 +1,13422 @@
+	{
+		flags.setOfflineErrorFlag();
+	}
+	inline bool isEmpty()
```
const

johnnovak commented on src/hardware/imf.cpp:
```diff
@@ -0,0 +1,13422 @@
+	{
+		return flags.isEmpty();
+	}
+	inline bool hasData()
```
const

johnnovak commented on src/hardware/imf.cpp:
```diff
@@ -0,0 +1,13422 @@
+	{
+		return flags.hasData();
+	}
+	inline bool hasError()
```
const

johnnovak commented on src/hardware/imf.cpp:
```diff
@@ -0,0 +1,13422 @@
+	{
+		return flags.hasError();
+	}
+	inline bool hasReadError()
```
const

johnnovak commented on src/hardware/imf.cpp:
```diff
@@ -0,0 +1,13422 @@
+	{
+		return flags.hasReadError();
+	}
+	inline bool hasOverflowError()
```
const

johnnovak commented on src/hardware/imf.cpp:
```diff
@@ -0,0 +1,13422 @@
+	{
+		return flags.hasOverflowError();
+	}
+	inline bool hasOfflineError()
```
const

johnnovak commented on src/hardware/imf.cpp:
```diff
@@ -0,0 +1,13422 @@
+	{
+		return flags.hasOfflineError();
+	}
+	inline uint8_t getFlagsByteValue()
```
const

johnnovak commented on src/hardware/imf.cpp:
```diff
@@ -0,0 +1,13422 @@
+enum SystemDataAvailability {
+	NO_DATA_AVAILABLE,
+	MIDI_DATA_AVAILABLE,
+	SYSTEM_DATA_AVAILABLE,
```
UpperCamelCase

kcgen replied:
Will do; not in the assembly.

johnnovak commented on src/hardware/imf.cpp:
```diff
@@ -0,0 +1,13422 @@
+static_assert(sizeof(MidiDataPacket) == 0x20,
+              "MidiDataPacket needs to be 0x20 in size!");
+
+static const std::string m_noteToString[12] =
```
Drop `m_` prefix.

johnnovak commented on src/hardware/imf.cpp:
```diff
@@ -0,0 +1,13422 @@
+	        : fraction(nf),
+	          note(nn)
+	{}
+	constexpr uint16_t getuint16_t() const noexcept
```
`GetUint16`?

FeralChild64 commented:
@kcgen Could you write what’s this? A music card emulation? Some 3rd party music patch for a single game? I have no idea what this patch is doing, and the size is really huge.

johnnovak commented on src/hardware/imf.cpp:
```diff
@@ -0,0 +1,13422 @@
+
+#pragma pack(push, 1)
+struct Duration {
+	uint16_t value = 0;
```
What is the unit of measurement for the duration? Can we encode it in the name?

johnnovak commented on src/hardware/imf.cpp:
```diff
@@ -0,0 +1,13422 @@
+	}
+	inline void setEmpty()
+	{
+		value |= 0x8000;
```
Extract duplicate 0x8000 constant and give it some name?

kcgen replied:
(not resolving because they  map to assembly literals)

johnnovak commented on src/hardware/imf.cpp:
```diff
@@ -0,0 +1,13422 @@
+static Duration ZERO_DURATION(0);
+
+#pragma pack(push, 1)
+struct PitchbenderValueMSB {
```
Maybe just call it `PitchbendValueMSB` 😄

johnnovak commented on src/hardware/imf.cpp:
```diff
@@ -0,0 +1,13422 @@
+	d   System Exclusive                    [F0], [F7]
+		System Common                       [F1], [F2], [F3], [F6]              [70]~[7F]
+	e   System Real-Time                    [F8], [FA]~[FC]
+	f   ??? only to SP                      N/A                                 N/A
```
Wonky indentation

kcgen replied:
I've protected a lot of block comments with `// clang-format off` to prevent it from wrapping and other things. Here's what happens if I take it off here:

![2023-03-28_01-39](https://user-images.githubusercontent.com/1557255/228179624-8c77a69d-da2f-4a36-8222-a130ac4514dd.png)

I've just manually formatted it to the margin now:

```c++

// clang-format off
/*
Format: <00fedcba>
Bit Affected Message                    Status Byte                         Status Byte (Event List)
--- ----------------------------------- ----------------------------------- --------------------------
a   Note ON/OFF                         [80]~[9F]                           [00]~[2F]
b   After Touch and Pitchbend           [A0]~[AF], [D0]~[DF], [E0]~[EF]     [50]~[5F], [60]~[6F]
c   Control Change                      [B0]~[BF]                           [30]~[3F]
	Program Change                      [C0]~[CD]                           [40]~[4F]
d   System Exclusive                    [F0], [F7]
	System Common                       [F1], [F2], [F3], [F6]              [70]~[7F]
e   System Real-Time                    [F8], [FA]~[FC]
f   ??? only to SP                      N/A                                 N/A

Note: Undefined MIDI status bytes ([F4], [F5], [F9], [FD], [FF]) and the active sensing code ([FE]) not listed
      in the preceding table are blocked. Succeeding data bytes are also blocked.
*/
// clang-format on
```

johnnovak commented on src/hardware/imf.cpp:
```diff
@@ -0,0 +1,13422 @@
+	// When CHAIN mode is cancelled, this path returns  to its previous state.
+	// Even if this path has been closed by entry into CHAIN mode, status
+	// reports from the music card reflect its programmed state.
+	uint8_t MidiIn_To_MidiOut = 0;
```
Add new lines for readability, remove underscores from names.

kcgen replied:
Fixed up whitespace, good catch @johnnovak - those variables were swallowed up around their descriptions.

ThomasEricB commented:
> @kcgen Could you write what’s this? A music card emulation? Some 3rd party music patch for a single game? I have no idea what this patch is doing, and the size is really huge.

@FeralChild64 Let me give @kcgen a hand. This is what ChatGPT's GPT's summary of all the commit names:

The pull request, titled "Add Loris Chiocca's IBM Music Feature patch #2367", integrates the IBM Music Feature (IMF) card support into the DOSBox Staging. It includes multiple commits that refactor and optimize the code, addressing various aspects such as integer types, explicit namespace usage, macro handling, initialization, const correctness, header names, loops, constructors, and more. Additionally, it incorporates changes to make the code conform to C++ standards and DOSBox Staging conventions. The commits also cover improvements in the code's functionality, such as configuring initial settings from byte streams, standardizing mixer rendering, and adding settings for IMFC IRQ, port, and filter.

FeralChild64 commented:
@ThomasEricB OK, now I understand. I didn’t realize „IBM Music Feature” is a sound card name, and Google search did not find anything useful :)

kcgen commented:
@johnnovak , 

The PR includes the clean-room reverse engineered assembly structures and functions (over 20k LOC, in the contrib/ ara). The assembly it annotated with the names that map over to the C++ code.

I think we can try a couple small style adjustments - but my main worry is that we break this relationship for anyone trying to work with the assembly in some systematic way.

Another option is that we put a note at the top about how the style is "per the original code".

kcgen commented:
## LGR Space Quest 3

https://www.youtube.com/watch?v=lyIBfjsTZcQ&t=144s

## PR Space Quest 3

https://user-images.githubusercontent.com/1557255/228140307-60cdca55-c70e-4cf6-a0b7-0e1bff74ec5d.mp4

ThomasEricB commented:
> LGR Space Quest 3: https://www.youtube.com/watch?v=lyIBfjsTZcQ&t=144s
> 
> PR's Space Quest 3 (w/ reverb and chorus enabled):

I've watched the video fully now. @kcgen have you seen this: https://github.com/Scalibq/FB-01-Emulator . Perhaps we want to emulate FB-01 Yamaha as well, since it has more support than the IBM Music Feature's handful of Sierra titles?

kcgen commented:
> FB-01

That's a great thread :+1: and to add to it, Scali has a very thorough article about his adventures in IMFC and FB-01 land, emulating the latter and included a video showing all the pieces in play:

https://scalibq.wordpress.com/2019/03/26/the-ibm-music-feature-card-and-yamaha-fb-01/

johnnovak commented:
> @johnnovak ,
> 
> The PR includes the clean-room reverse engineered assembly structures and functions (over 20k LOC, in the contrib/ area). The assembly is annotated with names that map over to the C++ code.
> 
> I think we can try a couple small style adjustments - but my main worry is that we break this relationship for anyone trying to work with the assembly in some systematic way.
> 
> Another option is that we put a note at the top about how the style is "per the original code".

Right, wasn't aware of that. In that case best to leave the names alone, 100%.

kcgen commented:
> This is a very nice addition, and Silpheed sounds really good on the card!

100%!

> I'm wondering if there's much point in continuing with my review because I can only comment on superficial stuff, and I agree we should keep the symbol names in sync with the asm code.

I've addressed almost all of them that don't hit the assembly.

Updated the branch; now switched over to `[imfc]` conf settings (and names).

Would appreciate your ears and see if we should apply a default filter. I think that's the big outstanding thing we can still improve before merging.

FeralChild64 commented on src/hardware/imfc_rom.c:
```diff
@@ -0,0 +1,1061 @@
+
```
Why do we need this file at all? It is not included in the compilation, `imfc.cpp` seems to contain the data too.

BTW, moving the binary data out of `imfc.cpp` is, IMHO, a good idea.

ThomasEricB replied:
Line 23 claims it features memory tables for the IBM Music Feature Card,

kcgen replied:
> BTW, moving the binary data out of imfc.cpp is, IMHO, a good idea.

Yes; it makes it a lot easier to read through the code. The `.c` file is  `#include`d just above the `class MusicFeatureCard` where they're used.

ThomasEricB reviewed (approved):
Sounds all good to me!

Grounded0 commented:
https://www.youtube.com/watch?v=lyIBfjsTZcQ

kcgen commented:
For those interested in matching the IMFC hardware recordings [shared by Pierre on YouTube](https://www.youtube.com/watch?v=Eoxkwy6a4R4&list=PLZ2JtZDT6cQI7Mo8mpl6xU1oDisbQN7k9), you can try setting `chorus = 80`. 

For example, here is Pierre's hardware recording:

https://github.com/dosbox-staging/dosbox-staging/assets/1557255/13a3e29f-ef3f-4ca6-a932-dac28f3ce79d

Here is DOSBox Staging 0.81 (alpha) with all defaults, plus `chorus = 80`:

https://github.com/dosbox-staging/dosbox-staging/assets/1557255/1951c5d8-6de5-4ce0-8304-827d6e0845fa

And here's the visual comparison of them:

![2023-10-23_11-37](https://github.com/dosbox-staging/dosbox-staging/assets/1557255/064d448b-645f-4ee7-8144-47e3365de1fd)
